LSP: Add support for call hierarchies

cbarrete opened the pull request:
This adds support for Call Hierarchies, coming in [version 3.16 of the LSP specification](https://github.com/microsoft/language-server-protocol/blob/gh-pages/_specifications/specification-3-16.md#textDocument_prepareCallHierarchy).

Commit:
LSP: Add support for call hierarchies

Commit:
LSP: Add support for call hierarchies

Commit:
LSP: Add support for call hierarchies

fsouza commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -136,6 +136,39 @@ function M.document_symbol()
+
+function M.incoming_calls()
+  local params = util.make_position_params()
+  request('textDocument/prepareCallHierarchy', params, function(_, _, result)
```
How will this work with multiple language servers?

cbarrete replied:
What do you mean exactly?

fsouza replied:
There may be multiple servers running for a given language. Currently neovim sends the request to all servers and you get a list of responses in the callback (most methods look at the first result), I think.

fsouza replied:
Sorry nevermind you're already handling it in the capability check (if the user has multiple servers with repeated capabilities it's kidna hard to know what to do hehe)

tjdevries replied:
Should this be a synchronous request?

cbarrete replied:
Probably not, getting the list of all calls can easily take over 10 seconds on large projects in CLion and it might be worse for some servers.

teto commented:
sounds cool. Know any server implenmenting this ?

cbarrete commented:
[rust-analyzer](https://github.com/rust-analyzer/rust-analyzer/) does, although [it still has some rough edges for now](https://github.com/rust-analyzer/rust-analyzer/issues/5103). I haven't found others yet unfortunately.

tjdevries commented:
Could you add a test for this? For example, if you can get a realistic looking message from rust-analyzer and then write a test against it, that'd be very helpful in getting this merged.

Seems pretty cool!

(Also, a gif/move/asciinema would be helpful to understand exactly what happens for this, if possible).

cbarrete commented:
Sure, I'll do both asap.

Commit:
LSP: Jump to call location

Jump to the call site instead of jumping to the definition of the
caller/callee.

cbarrete commented:
[Here](https://asciinema.org/a/bNaIqeRSUT2lKiXjeg6N95FPV) is an asciinema, I'll add a test next.

tjdevries commented:
(If you need help writing the test, let me know!)

cbarrete commented:
I haven't had the time to look into it yet. Are there utilities for checking the contents of the quickfix window?

tjdevries commented:
You could just `getqflist` ? which would be `vim.fn.getqflist` from lua

Commit:
LSP: add tests for the call hierarchy callbacks

tjdevries commented:
Hey @cbarrete, this looks great! I think the only thing stopping this from getting merged is that you need to run `make lint` locally and fix the lint errors. The tests look good! Thanks for adding them.

I'm assuming there's nothing left you wanted to do for this PR

Commit:
Fix linting error

cbarrete commented:
Thanks for letting me know, it should be good now!

tjdevries commented:
Cool, assuming tests pass now, I will merge! Thanks for the hard work!
